File: src/main/kotlin/si/pecan/dto/ChatRoom.kt
package si.pecan.dto

import si.pecan.model.InstantMessage
import java.time.LocalDateTime
import java.util.*


/**
 * Created by m.pecan on 22/03/2017.
 */
data class ChatRoom(val id: UUID, val initiator: User, val target: User, val messages: List<Message>, val created: LocalDateTime, val lastMessageTime: LocalDateTime?)

data class Message(val poster: String, val content: String, val created: LocalDateTime, val id: UUID, var chatId: UUID? = null)

fun InstantMessage.toDto() = Message(postedBy.username, content, created, this.id!!)

File: src/react/js/api/chat_room.js
import {User} from './user';
import {Map} from 'immutable'
import moment from 'moment';
export class ChatRoom {
    constructor(response) {
        this.id = response.id;
        this.initiator = new User(response.initiator);
        this.target = new User(response.target);
        this.messages = Map({}).merge(response.messages
            .map((message) => {
                message.created = moment(message.created);
                return message;
            })
            .reduce((agg, val) => {
                agg[val.id] = val;
                return agg
            }, {}));
        this.created = response.created;
    }
}